core/src/chain/mod.rs

```diff
@@ -131,82 +204,109 @@ pub trait SourceChain {
     /// Pair to ensure the chain links up correctly across the underlying table data
     /// the newly created and pushed Pair is returned in the fn Result
-    fn push_entry(&mut self, entry: &Entry) -> Result<Pair, HolochainError>;
+    fn commit_entry(&mut self, entry: &Entry) -> Result<Pair, HolochainError>;
     /// get an Entry by Entry key from the HashTable if it exists
     fn entry(&self, entry_hash: &str) -> Result<Option<Pair>, HolochainError>;
 
     /// pair-oriented version of push_entry()
-    fn push_pair(&mut self, pair: &Pair) -> Result<Pair, HolochainError>;
-    /// get a Pair by Pair/Header key from the HashTable if it exists
-    fn pair(&self, message: &str) -> Result<Option<Pair>, HolochainError>;
+    fn commit_pair(&mut self, pair: &Pair) -> Result<Pair, HolochainError>;
+
+    // /// get a Pair by Pair/Header key from the HashTable if it exists
+    fn pair(&self, pair_hast: &str) -> Result<Option<Pair>, HolochainError>;
 }
 
 impl SourceChain for Chain {
     fn top_pair(&self) -> Option<Pair> {
-        self.actor.top_pair()
+        self.chain_actor.top_pair()
     }
 
     fn set_top_pair(&self, pair: &Option<Pair>) -> Result<Option<Pair>, HolochainError> {
-        self.actor.set_top_pair(&pair)
+        self.chain_actor.set_top_pair(&pair)
     }
 
-    fn top_pair_type(&self, t: &str) -> Option<Pair> {
+    fn top_pair_of_type(&self, t: &str) -> Option<Pair> {
         self.iter().find(|p| p.header().entry_type() == t)
     }
 
-    fn push_pair(&mut self, pair: &Pair) -> Result<Pair, HolochainError> {
+    /// Whole process of authoring an entry.
+    /// 1. `validation` of the new entry using the ribosome and validation WASM code
+    /// 2. `pushing` the new entry onto the source chain, if valid
+    /// 3. `putting` the entry into the (distributed) hash table, if defined as public
+    fn commit_pair(&mut self, pair: &Pair) -> Result<Pair, HolochainError> {
+
+        // 1. validation
         if !(pair.validate()) {
             return Err(HolochainError::new(
                 "attempted to push an invalid pair for this chain",
             ));
         }
 
         let top_pair = self.top_pair().as_ref().map(|p| p.key());
-        let next_pair = pair.header().link();
+        let prev_pair = pair.header().link();
 
-        if top_pair != next_pair {
+        if top_pair != prev_pair {
             return Err(HolochainError::new(&format!(
-                "top pair did not match next hash pair from pushed pair: {:?} vs. {:?}",
-                top_pair, next_pair,
+                "top pair did not match previous hash pair from commited pair: {:?} vs. {:?}",
+                top_pair, prev_pair,
             )));
         }
 
-        self.table.commit(&pair.clone())?;
+        // 2. pushing
+        // 3. putting
+        let header_entry = &pair.clone().header().to_entry();
+        // println!("Chain.commit_pair() header_entry = {:?}", header_entry);
+        self.table_actor.put(header_entry)?;
```

Comment:
For iterating through the chain, header by header. See `ChainIterator.next()`
How am I suppose to do that otherwise?

Reply:
@lucksus the header has been in the HT literally always

Reply:
@ddd-mtl: Why does the HT iterate through the chain??

@thedavidmeister: Just because it was that way before (which I pointed out as not fitting the mental model already) doesn't make it right, right?. ;) 

I don't see why the HT should deal with headers from a logical stand point and I would like to understand if there is an implementation reason for it. I am concerned that we produce more and more code that reproduces this wrong pattern. Like this line which gets added in here. Although HT deals with entries now (in put and get), it still needs to have a header entry putted to it explicitly? Why would you do that? My feeling is that we are building something here that is not really the thing we need..